### PR TITLE
Replace custom hash table with std::unordered_map

### PR DESCRIPTION
**Pull request by mitrokosta**
I replaced the "HASHLINE" in variables and function tables with std::unordered_map.

I rewrote both tables from scratch and I'm not sure if my naming code style was correct. I tried to conform to Google code style, because it was the most voted one.

Requires a small script change: in [this](https://github.com/storm-devs/sd-teho-public/blob/e7d161eb1d4da25a73a517ff6d8373d8acbb6609/program/characters/LSC_Q2Utilite.c#L1352) line a variable with an empty name is declared which normally should be a compile error. Just need to remove the comma and everything will compile correctly.

### CODE REVIEW

**Comment by espkk** [comment]
Thanks for the PR! That is something that had to be done for sure.

The only thing I'm a bit concerned about is that boost/std::unordered_list is relatively slow in comparison to some custom (e.g. google's) hashtable implementions. I did some measurements a while ago specifically for SE but it seems that I lost everything :(

But anyway, switching to the std from self-made hashtable already should be winning in terms of performance and maintainability.

I will look though it soon

File: src/libs/Common_h/strutils.h
Comment (np-c0mp1ete) [inline src/libs/Common_h/strutils.h]: there is also a function for case-insensitive string comparison from @Hammie in https://github.com/storm-devs/storm-engine/blob/develop/src/libs/util/include/storm/string_compare.hpp

could it be reused/unified? if not, maybe at least put them in a single file?

File: src/apps/ENGINE/s_vartab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_vartab.cpp]: maybe log a warning here?

File: src/apps/ENGINE/s_vartab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_vartab.cpp]: log a warning? or maybe return a bool and log a warning on the caller site if result != true?

File: src/apps/ENGINE/s_vartab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_vartab.cpp:94]: would be nice to handle a case when a given segment_id is not found

File: src/apps/ENGINE/s_functab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_functab.cpp:96]: handle a case when `segment_id` is not found?

File: src/apps/ENGINE/s_functab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_functab.cpp]: return an error when `func_index >= funcs_.size()`?

File: src/apps/ENGINE/s_functab.cpp
Comment (np-c0mp1ete) [inline src/apps/ENGINE/s_functab.cpp]: return an error when `func_index >= funcs_.size()`?

**Review by np-c0mp1ete — COMMENTED** [review]
thanks for the PR! 
I think it's better to stick to the standard library unless it's found to be a performance bottleneck. Switching to abseil or another lib shouldn't be hard if needed.

File: src/libs/Common_h/strutils.h
Comment (Hammie) [reply src/libs/Common_h/strutils.h]: It should do the same, and you could also update mine to use `std::equal` in the process ;)

File: src/apps/ENGINE/s_vartab.h
Comment (Hammie) [inline src/apps/ENGINE/s_vartab.h]: Is this shared_ptr used anywhere else? Or could it be a unique_ptr?

Comment (mitrokosta) [reply src/apps/ENGINE/s_vartab.h]: With the current interface of VarTable it can't be replaced with a unique_ptr. VarTable::GetVar and VarTable::GetVarX require to make a copy of a VarInfo.

Comment (espkk) [reply src/apps/ENGINE/s_vartab.h]: I have not had a chance to review the code yet, but I believe we definitely shall try to avoid using std::shared_ptr there
It would be a significant drop in performance due to a control block
I will try to check this soon and propose a solution

Comment (mitrokosta) [reply src/apps/ENGINE/s_vartab.h]: I rewrote the code, so it uses a const pointer to VarInfo instead of a copy. This allowed me to get rid of shared_ptr. Let me know if this is a good solution.

File: src/apps/ENGINE/s_vartab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_vartab.cpp:94]: How do you suggest to handle it? It is not an error - a segment may have no variables at all, theoretically (if it only operates with global variables from another segment, for example).

File: src/apps/ENGINE/s_functab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_functab.cpp:96]: Same as above, some segments may have no functions at all, only variables.

File: src/apps/ENGINE/s_functab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_functab.cpp]: Done.

File: src/apps/ENGINE/s_functab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_functab.cpp]: Done.

File: src/apps/ENGINE/s_vartab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_vartab.cpp]: Done.

File: src/apps/ENGINE/s_vartab.cpp
Comment (mitrokosta) [reply src/apps/ENGINE/s_vartab.cpp]: Done.

File: src/libs/Common_h/strutils.h
Comment (mitrokosta) [reply src/libs/Common_h/strutils.h]: I moved my hasher and comparator to this file, also reused @Hammie 's iEquals function. Also I updated iEquals to use std::equal.

File: src/apps/ENGINE/s_vartab.cpp
Comment (np-c0mp1ete) [reply src/apps/ENGINE/s_vartab.cpp:94]: I see. I guess it's fine to leave it as it is then.

**Review by np-c0mp1ete — DISMISSED** [review]
Thanks for the PR! I have no further comments, but I'll let @espkk to review it before merging too.

File: src/apps/ENGINE/s_functab.h
Comment (espkk) [inline src/apps/ENGINE/s_functab.h:29]: can you explain how it is used? I cannot find it's by inspecting the commit...

Comment (mitrokosta) [reply src/apps/ENGINE/s_functab.h:29]: It is used in FuncTable::FindVar, [this](https://github.com/storm-devs/storm-engine/blob/2bea837a898bc16d6209318065fd48a2cdabc1bc/src/apps/ENGINE/s_functab.cpp#L178) line. The same thing was done in the old code.

Comment (espkk) [reply src/apps/ENGINE/s_functab.h:29]: So it is hashed twice, right? I recall now that this was another point against stock unordered_map
Though it is still fine for now
Everything else is good! I think this can be merged after rebase

Comment (mitrokosta) [reply src/apps/ENGINE/s_functab.h:29]: Variable names in FuncTable are not actually hashed twice, because they are not stored in an unordered_map. They are stored in a vector (local_vars) which is stored in FuncInfo which is stored in an unordered_map. In order to find a functions' argument you have to look that function up first, using unordered_map, and then find the argument in local_vars vector, using this argument's hash to speed up comparison. The same thing was done in the old code, the difference is that it used a custom hash table.

Comment (espkk) [reply src/apps/ENGINE/s_functab.h:29]: Ah, that makes sense
I believe I tried to invent a container for string interning. I see now this is unchanged here

**Review by espkk — APPROVED** [review]

**Pull request merged**
